File: RetClass.py

class Rectangles:
    x=0
    y=0
    w=0
    h=0
    name="name"
    
    def __init__(self,x,y,w,h,retname):
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.name = retname
        
    def origin(self):
        return [self.x,self.y]
        
    def w(self):
        return self.w
        
    def h(self):
        return self.h
        
        
    def getRectPointsX(self):
        return list([self.x,self.x+self.w,self.x+self.w,self.x,self.x])
    def getRectPointsY(self):
        return list([self.y,self.y,self.y-self.h,self.y-self.h,self.y])
        
    def IntersectionWith(self,ret):
    
        if(ret.name != self.name):
            retpointsx = ret.getRectPointsX()
            retpointsy = ret.getRectPointsY()
        
            for i in range(0,len(retpointsx)):
                if((retpointsx[i]>=self.x and retpointsx[i]<=self.x+self.w) and (retpointsy[i]<=self.y and retpointsy[i]>=self.y-self.h)): return True
            
            
            retpointsx = self.getRectPointsX()
            retpointsy = self.getRectPointsY()
        
            for i in range(0,len(retpointsx)):
                if((retpointsx[i]>=ret.x and retpointsx[i]<=ret.x+ret.w) and (retpointsy[i]<=ret.y and retpointsy[i]>=ret.y-ret.h)): return True
        
        return False

File: retclass.cpp
// Rectangles.h
#ifndef RECTANGLES_H
#define RECTANGLES_H

#include <string>
#include <iostream>
#include <vector>

class Rectangles {
private:
    double x;
    double y;
    double w;
    double h;
    std::string name;

public:
    Rectangles(double x, double y, double w, double h, std::string retname) 
        : x(x), y(y), w(w), h(h), name(retname) {}

    std::vector<double> origin() {
        return {x, y};
    }

    double getWidth() {
        return w;
    }

    double getHeight() {
        return h;
    }

    std::vector<double> getRectPointsX() {
        return {x, x + w, x + w, x, x};
    }

    std::vector<double> getRectPointsY() {
        return {y, y, y - h, y - h, y};
    }

    bool IntersectionWith(Rectangles ret) {
        if (ret.name != name) {
            std::vector<double> retpointsx = ret.getRectPointsX();
            std::vector<double> retpointsy = ret.getRectPointsY();

            for (size_t i = 0; i < retpointsx.size(); ++i) {
                if ((retpointsx[i] >= x && retpointsx[i] <= x + w) &&
                    (retpointsy[i] <= y && retpointsy[i] >= y - h)) {
                    return true;
                }
            }

            retpointsx = getRectPointsX();
            retpointsy = getRectPointsY();

            for (size_t i = 0; i < retpointsx.size(); ++i) {
                if ((retpointsx[i] >= ret.x && retpointsx[i] <= ret.x + ret.w) &&
                    (retpointsy[i] <= ret.y && retpointsy[i] >= ret.y - ret.h)) {
                    return true;
                }
            }
        }

        return false;
    }
};

#endif // RECTANGLES_H
